src/main/scala/format/pgn/Parser.scala

```diff
@@ -98,296 +63,200 @@ object Parser {
 
-    def moveExtras: Parser[Unit] =
-      as("moveExtras") {
-        commentary.^^^(())
-      }
+    val commentary = (blockCommentary | inlineCommentary).withContext("Invalid comment") <* whitespaces
 
-    def nagGlyphs: Parser[Glyphs] =
-      as("nagGlyphs") {
-        rep(nag) ^^ { nags =>
-          Glyphs fromList nags.flatMap { Glyph.find _ }
-        }
-      }
+    val result: P[String] = P.stringIn(List("*", "1/2-1/2", "½-½", "0-1", "1-0"))
 
-    val nagGlyphsRE = Glyph.PositionAssessment.all
-      .map(_.symbol)
-      .sortBy(-_.length)
-      .map(Regex.quote(_))
-      .mkString("|")
-      .r
+    val nagGlyphsRE = P.stringIn(
+      Glyph.PositionAssessment.all
+        .map(_.symbol)
+        .sortBy(-_.length)
+    )
 
-    def nag: Parser[String] =
-      as("nag") {
-        """\$\d+""".r | nagGlyphsRE
-      }
+    val nag = (P.char('$') ~ R.digit.rep).string | nagGlyphsRE
 
-    def variation: Parser[List[StrMove]] =
-      as("variation") {
-        "(" ~> strMoves <~ ")" ^^ { case (_, sms, _) => sms }
+    val nagGlyphs: P0[Glyphs] = (nag <* whitespaces).rep0.map(nags =>
+      Glyphs fromList nags.flatMap {
+        Glyph.find
       }
+    )
+
+    val moveExtras = commentary.as(())
+
+    val positiveIntString: P[String] =
+      (N.nonZeroDigit ~ N.digits0).string
+
+    // '. ' or '... ' or '. ... '
+    val numberSuffix = (P.char('.') | whitespace).rep0.void
+
+    // 10. or 10... but not 0 or 1-0 or 1/2
+    val number = (positiveIntString <* !P.charIn('-', '/') ~ numberSuffix).string
+
+    val forbidNullMove =
+      P.stringIn(List("--", "Z0", "null", "pass", "@@@@"))
+        .?
+        .flatMap(o => o.fold(P.unit)(_ => P.failWith("Lichess does not support null moves").void))
 
-    def commentary: Parser[String] = blockCommentary | inlineCommentary
+    val strMove: P[San] = P
+      .recursive[San] { recuse =>
+        val variation: P[Sans] =
+          (((P.char('(') <* whitespaces) *> recuse.rep0 <* (P.char(')') ~ whitespaces)) <* whitespaces)
+            .map(Sans(_))
 
-    def blockCommentary: Parser[String] =
-      as("block comment") {
-        "{" ~> """[^\}]*""".r <~ "}"
+        ((number.backtrack | commentary <* whitespaces).rep0 ~ forbidNullMove).with1 *>
+          (((MoveParser.moveWithSuffix ~ nagGlyphs ~ commentary.rep0 ~ nagGlyphs ~ variation.rep0) <* moveExtras.rep0) <* whitespaces).backtrack
+            .map { case ((((san, glyphs), comments), glyphs2), variations) =>
+              san withComments comments withVariations variations mergeGlyphs (glyphs merge glyphs2)
+            }
       }
 
-    def inlineCommentary: Parser[String] =
-      as("inline comment") {
-        ";" ~> """.+""".r
+    val strMoves: P[(InitialPosition, List[San], Option[String])] =
+      ((commentary.rep0.with1 ~ strMove.rep) ~ (result <* whitespaces).? ~ commentary.rep0).map {
+        case (((coms, sans), res), _) => (InitialPosition(cleanComments(coms)), sans.toList, res)
```

Comment:
looks like you are ignoring the `commentary.rep0` so you may get better performance using `<* commentary.rep0` or equivalently `~ commentary.rep0.void`